### PR TITLE
Fix TS_ASSERTS to be threadsafe

### PR DESCRIPTION
**Pull request by ylow**
An annoying side effect is that TS_ASSERTS are now *really* slow.
In particular for some tests that loop TS_ASSERTS millions of times,
this will degrade test performance significantly. Those tests may have
to change from:

  TS_ASSERT(something);

to

  if (something) {
    TS_ASSERT(something);
  }

It will also be problematic if TS_ASSERT blocks, in particular, in
tests of threading primitives. For instance this is not a good idea.

  TS_ASSERT(wait_for_thread());

instead:

  int ret = wait_for_thread();
  TS_ASSERT(ret);

### CODE REVIEW

File: src/flexible_type/flexible_type.hpp [hunk 1]
```diff
@@ -445,6 +445,33 @@ class flexible_type {
   template <typename T>
   const T& get() const;
 
+
+  /**
```
Comment (znation) [inline src/flexible_type/flexible_type.hpp]: This part doesn't look related to making TS_ASSERT thread safe. Should there be two separate commits here?

Comment (ylow) [reply src/flexible_type/flexible_type.hpp]: Ah yes. indeed. This also fixes a bunch of other test failures. I can rename the PR. Or I can submit that as a different PR. What do you prefer?

Comment (ylow) [reply src/flexible_type/flexible_type.hpp]: I think it I will split into a different PR.

**Comment by ylow** [comment]
PR has been split.

**Comment by hoytak** [comment]
Why don't we just format all our ASSERT_* statements so they can be picked up by the rest of our test system, and just use those here?

**Comment by ylow** [comment]
ASSERTs just raise exceptions. TS_ASSERTS bind into the test harness system (boost test in this case)
